features/323-login-id/design.md

```diff
@@ -0,0 +1,109 @@
+6. To signup using password authentication, user can provide a list of tuple (login ID key, login ID). e.g. `{ "username": "test", "email": "a@example.com" }`
+7. For password authentication principals, user can login using a tuple (login ID key, login ID). e.g. `{ "email": "a@example.com" }`
+8. Login IDs are globally unique in app, without considering associated login ID keys.
+9. Auth gear does not knows the semantics of login IDs.
+
+
+## Problem
+1. Some features requires semantics of login IDs to function.
+2. Some authentication system designs may not be implemented easily in current design.
+
+
+# Use cases
+
+Considering some practical scenarios of authentication system:
+
+## GitHub
+- Supports multiple emails per user, with email verification.
+- User can add/delete any amount of emails.
+- Only verified emails would be associated with the user in public.
+- User can allow recovery email/security alert to be sent to all verified emails or a specific verified email.
+
+
+## Facebook
+- User can login with email/phone number/username, without indicating what type of value it is.
+- User can add/delete any amount of emails, which can be used to login after verification.
+- Disposable email address is blocked.
+- All emails/phones of user is verified.
+- Email communications are personalized with user language & name
+
+
+# Design Objectives
+- Allow auth related features to understand semantics of some login IDs / user metadata
+- Support multiple login ID with same login ID key
+- Support adding/removing login ID
+
+
+# Proposed Design
+
+## Use Principals in Auth API
+When login:
+- Specify which identity provider and principal identifier is used to authenticate.
+  For password authentication, there is no need to specify login ID key (login ID is already globally unique).
+  If identity provider is not specified, password identity provider is used by default.
+  Example:
+```json
+{
+    "provider": "password",
+    "login_id": "a@example.com",
+    "password": "12345678"
+}
+{
+    "provider": "jwt",
+    "token": "..."
+}
+```
+
+- If login succeed, the principal would be associated with current access token and returned:
+```json
+{
+    "access_token": "...",
+    "user": {},
+    "principal": { "email": "a@example.com" }
+}
+```
+
+
+## Dedicated Principal Manipulation API
+Dedicated APIs should be provided for principal manipulations, considering:
+- User may have a lot of principals (due to unlimited login emails/external identity providers/etc.)
+- User may create principals with limited permission (for automatic process/etc.)
+- User may want to add/remove login emails
+
+On sign up, only single principal is allowed to be created along with the new user.
+User can add/remove principals after signing up (e.g. login email)
+
+
+## Standard Semantics for Login ID Key & User Metadata
+With reference to OpenID Connect standard claims,
+a list of standard login ID key & user metadata would be specified,
```

Comment:
I think it is similar to the idea of in future, metadata will have some standardized fields like avatar, that other part of Skygear will respect and use it semantically?